File: 0169-majority-element/0169-majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
               // unordered_map<int,int>map;
        // for(int i=0;i<nums.size();i++){
        //     map[nums[i]]++;
        // }
        // for(auto num : map){
        //   if(num.second>nums.size()/2){
        //     return num.first;
        //   }
        // } 
        // return -1 ;
        // time complexity o(n);
        //space complexity o(n)

        //Another method.
        //More's voting method for mojority;
        int n=nums.size();
        int count=0;
        int ele=-1;
        for(int i=0;i<n;i++){
            if(count==0){
                ele=nums[i];
                count=1;
            }
            else if(ele==nums[i])count++;
            else count--;
        }
        count=0;
        for(int num:nums){
            if(num==ele)count++;
        }
        if(count>n/2)return ele;
        return -1;
    }
};

File: 0169-majority-element/0169-majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        double ele=1e9;
        int n=nums.length;
        int count=0;
        for(int i=0;i<n;i++){
            if(count==0){
                ele=nums[i];
                count++;
            }
            else if(ele!=nums[i]){
                count--;
            }
            else count++;
        }
        return (int)ele;
    }
}





// class Solution {
//     public int majorityElement(int[] nums) {
//         int n=nums.length;;
//         HashMap<Integer,Integer>map1=new HashMap<>();
//         for(int i=0;i<n;i++){
//             if(map1.containsKey(nums[i])){
//                 int count=map1.get(nums[i]);
//                 count++;
//                 map1.put(nums[i],count);
//             }
//             else map1.put(nums[i],1);
//         }
//         for(Map.Entry<Integer,Integer>entry:map1.entrySet()){
//             int key=entry.getKey();
//             int val=entry.getValue();
//             if(val>n/2)return key;
//         }
//          return -1;
//     }
// }